tools/crm_mon.c

```diff
@@ -952,8 +957,9 @@ pacemakerd_status(void)
                 if (rc == pcmk_rc_ok) {
                     pcmk_dispatch_ipc(pacemakerd_api);
                     rc = ENOTCONN;
-                    if ((output_format == mon_output_console) ||
-                        (output_format == mon_output_plain)) {
+                    if (options.one_shot
+                        || (output_format == mon_output_console)) {
+
```

Comment:
yeah this block is a good example, there shouldn't be an if/else here, just calling appropriate message functions (in this case probably a single message function that takes `state`)

that would also let us indicate this in xml format somehow

Reply:
If we even want to indicate it in non-console formats, sure. It depends on whether we think these should be transient messages like "Reconnecting...". If so, then they'd be console-only (plus `log` at the debug level) via `out->transient()`, and we wouldn't need messages configured specifically for this task.

I'm on the fence. I lean toward making them all transient, as I don't see much of a reward to be gained from calling a message here, and in general we should progress to a new state quickly. If we do go with messages, maybe we can repurpose the `pacemakerd-health` message functions.

Especially since so far, I haven't been able to get into these states to get any of these to actually print from `pacemakerd_status()` using any format or `crm_mon` run option. So maybe it's pretty rare that they would even appear?

I've got a draft of a message approach ready to go, but for the reasons discussed above, it doesn't seem to make a difference.

Reply:
We can do messages for the "Cluster is not available" message, regardless of what we decide for `pacemakerd_status()`.

Reply:
> For console, the ideal is that if everything proceeded in slow motion, the user would see a succession of messages indicating the start-up/shutdown progress.

That makes sense. That doesn't happen currently.

I added a sleep after each `pacemakerd_state` assignment and before the first assignment. In console mode, I get nothing but `"Waiting until cluster is available on this node ..."` when we're waiting for the first connection, and `"Connection to the cluster-daemons terminated"` followed by `"Reconnecting ..."` when we lost our connection (e.g., due to stopped cluster) and we're waiting to reconnect. We don't see any of the messages from `pacemakerd_status()`. It's a similar story in one-shot mode.

It might be due to the way resets are ordered relative to the `pacemakerd_status()` messages. I'll look more closely soon. Just saying we don't have to worry specifically about preserving the current display of those messages, since they don't display.

Originally I hoped to just use `out->transient()` calls. If we want to show these states in XML output too, then transient's not gonna work. It might not end up making sense to add the `pcmk__output_t:transient()` method in the first place, depending on what we decide to do with the `Waiting until...` and `Reconnecting...` messages (the first intended use case for `transient`).

Reply:
> > It could even be a blanket "cluster status not yet available" message or something like that.
> 
> We currently do get a generic message like that, for one-shot or console mode. For one-shot, you get `"crm_mon: Connection to cluster failed: Resource temporarily unavailable"`. For console, you get `"Waiting until cluster is available on this node ..."` if you haven't established a connection yet and `"Reconnecting ..."` if you have. Daemonized mode is the only case where you don't get a generic "cluster not available" type of message yet.

That's not bad, though it would be ideal for daemonized HTML/CGI to say something useful.

> 
> > For console, the ideal is that if everything proceeded in slow motion, the user would see a succession of messages indicating the start-up/shutdown progress.
> 
> That makes sense. I'm not sure that happens currently. It looks like even if you can catch the usually-ephemeral intermediate `pacemakerd` states, the `Waiting until cluster is available on this node ...` message or the `Reconnecting ...` message will clear the screen and overwrite the state before anyone observes it. One-shot mode might behave better if you can time it right, not sure.

Bah. Maybe pacemakerd_status() could give back a string that could be tacked on to "Waiting until ..." e.g. "(Pacemaker daemons starting)".

> 
> Adding sleeps in `pacemakerd` turns out not to work because it causes `pcmk_ipc_connect()` calls to hang in `pacemakerd_status()`. So while we're sleeping, we can't get the `pacemakerd` state. Delaying the state updates by doing it after some interval through the mainloop instead of synchronously would probably get around that.

Doh, of course. That may be more effort than this is worth.


> Gotta decide what to do with the `"Waiting until..."` and `"Reconnecting..."` messages if we have a specific `pacemakerd` state to display (considering the `clear()` calls and such)... and other than that, it might be a straightforward addition of a new message for each output format. We might be able to reuse the `pacemakerd-health` message for this, but it would require some hackiness:
> 
>     * text mode prints fields we don't want unless we're in quiet mode
> 
>     * console mode clears the screen which would mess with the display of transient messages like `"Reconnecting"`
> 
> 
> So a semi-redundant new message may be more appropriate, we'll see
> 
> Also thinking we could merge the `pacemakerd_status()` definitions from `crm_mon.c` and `pcmk_status.c` in some way

I don't remember why those are separate, I think it had something to do with 1-shot vs not

Reply:
> I don't remember why those are separate, I think it had something to do with 1-shot vs not

The one in `crm_mon.c` was added first. The one in `pcmk_status.c` was added later, but at a glance they seem to be doing the same thing with minor differences.
* `pcmk_status.c` doesn't do the conditional printing based on output format.
* `pcmk_status.c` uses `EAGAIN` instead of `ENOTCONN` as its default return code.
* `crm_mon.c` uses `dispatch_poll` while `pcmk_status.c` uses `dispatch_sync`.
* `crm_mon.c` returns `pcmk_rc_ok` early if `!use_cib_native`.
* `pcmk_status.c` prints a "Could not connect" message if `pcmk_ipc_connect()` fails

Might be able to make them both happy with a common internal function